src/MatrixTwitter.js

```diff
@@ -0,0 +1,987 @@
+                  },
+                  "state_key": ""
+              }
+          ]
+        }
+      }
+    );
+  });
+}
+
+MatrixTwitter.prototype._process_incoming_dm = function(msg){
+  var users = [msg.sender_id_str,msg.recipient_id_str].sort().join('');
+
+  this._update_user_timeline_profile(msg.sender);
+  this._update_user_timeline_profile(msg.recipient);
+
+  if(this.sent_dms.contains(users,msg.text)){
+    log.info("Twitter.DM","DM has already been processed, ignoring.");
+    return;
+  }
+
+  this.storage.get_dm_room(users).then(room_id =>{
+    if(room_id){
+      this._put_dm_in_room(room_id,msg);
+      return;
+    }
+    //Create a new room.
+    return this._create_dm_room(msg,users).then(room => {
+      this.storage.add_dm_room(room.room_id,users);
+      var mroom = new MatrixRoom(room.room_id);
+      var rroom = new RemoteRoom("dm_"+users);
+      rroom.set("twitter_type","dm");
+
+      this._bridge.getRoomStore().linkRooms(mroom,rroom);
+
+      this._put_dm_in_room(room.room_id,msg);
+    });
+  }).catch(reason =>{
+    log.error("Twitter.DM","Couldn't create room. Reason: " + reason);
+  });
+
+}
+
+MatrixTwitter.prototype._put_dm_in_room = function(room_id,msg){
+  var intent = this._get_intent(msg.sender_id_str);
+  log.info("Twitter.DM","Recieved DM from %s(%s) => %s(%s)",msg.sender_id_str,msg.sender_screen_name,msg.recipient_id_str,msg.recipient_screen_name);
+  intent.sendMessage(room_id, {"msgtype":"m.text","body":msg.text});
+}
+
+
+/**
+ * MatrixTwitter.prototype.send_dm - Send a DM on the users behalf. The room_id
+ * should be a DM room which has been set up for the user in advance.
+ *
+ * @param  {string} user_id    The user trying to send the message.
+ * @param  {string} room_id    The DM room that the message was sent from.
+ * @param  {string} text       The body text of the message.
+ * @return {Promise}           A promise that will resolve when the operation
+ * completes
+ */
+MatrixTwitter.prototype.send_dm = function(user_id,room_id,text){
+  //Get the users from the room
+  var users = "";
+  return this.storage.get_users_from_dm_room(room_id).then(u =>{
+    users = u;
+    if(users == null){
+      log.error("Twitter.DM","User (%s) tried to send a DM to (%s) but the room was not found in the DB. This shouldn't happen.",user_id,room_id);
+    }
+    return this._get_twitter_client(user_id);
+  }).then(client => {
+    var otheruser = users.replace(client.profile.id_str,"");
+    log.info("Twitter.DM","Sending DM from %s(%s) => %s",client.profile.id_str,client.profile.screen_name,otheruser);
+    this.sent_dms.push(users,text);
+    client.post("direct_messages/new",{user_id:otheruser,text:text},(error,response) =>{
+      if(error){
+        log.error("Twitter.DM","direct_messages/new failed");
+        console.log(error);
+      }
+    });
+  }).catch(reason =>{
+    log.error("Twitter.DM","Failed to send DM: %s", reason);
+  });
+}
+
+module.exports = {
+    MatrixTwitter: MatrixTwitter
+}
```

Comment:
When a file starts approaching 1000 lines.... you know you should probably look at it and see if it can be broken down into more manageable chunks. It's been horrible reviewing this file because there's so many functions that are flying around the place which can be called. You have a nice split between "hashtag" and "timeline" processing: that might be a good partition point.